### PR TITLE
Better assert raise messages

### LINKED ISSUE #3703: ExUnit.assert_raise: split mismatch message into serveral lines
Currently, assert_raise, shows the mismatch message, in one long line. such as:

```
  2) test cleans dependencies (Mix.Tasks.DepsTest)
     test/mix/tasks/deps_test.exs:570
     Wrong message for Mix.Error. Expected "mix deps.clean expects dependencies as arguments or a flag indicating which dependencies to clean. The --all option will clean all dependencies while the --unused option cleans unused dependencies", got "`mix deps.clean` expects dependencies as arguments or a flag indicating which dependencies to clean. The '--all' option will clean all dependencies while the '--unused' option cleans unused dependencies"
     stacktrace:
       test/mix/tasks/deps_test.exs:583: anonymous fn/0 in Mix.Tasks.DepsTest.test cleans dependencies/1
       (elixir) lib/file.ex:1131: File.cd!/2
       test/test_helper.exs:102: MixTest.Case.in_fixture/3
       test/mix/tasks/deps_test.exs:573
```

What I think it would be better to have is the code results split in lines,
and to have Expected and Received aligned, so it's easier to spot where there is a difference.

```
  2) test cleans dependencies (Mix.Tasks.DepsTest)
     test/mix/tasks/deps_test.exs:570
     Wrong message for Mix.Error.
     Expected:
       "mix deps.clean expects dependencies as arguments or a flag indicating which dependencies to clean. The --all option will clean all dependencies while the --unused option cleans unused dependencies"
     Got:
       "`mix deps.clean` expects dependencies as arguments or a flag indicating which dependencies to clean. The '--all' option will clean all dependencies while the '--unused' option cleans unused dependencies"
     stacktrace:
       test/mix/tasks/deps_test.exs:583: anonymous fn/0 in Mix.Tasks.DepsTest.test cleans dependencies/1
       (elixir) lib/file.ex:1131: File.cd!/2
       test/test_helper.exs:102: MixTest.Case.in_fixture/3
       test/mix/tasks/deps_test.exs:573
```

### PR DESCRIPTION
**Pull request by whatyouhide**
This should close #3703. Also, I added tests for `assert_raise/3` because there were none :).

### CODE REVIEW

File: lib/ex_unit/lib/ex_unit/assertions.ex [hunk 1]
```diff
@@ -392,8 +392,13 @@ defmodule ExUnit.Assertions do
...
 
-    msg = "Wrong message for #{inspect exception}. " <>
-          "Expected #{inspect message}, got #{inspect Exception.message(error)}"
+    msg = """
+    Wrong message for #{inspect exception}
+    Expected:
+      #{inspect message}
+    Got:
+      #{inspect Exception.message(error)}
+    """
```
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/assertions.ex]: This adds extra newline at the end.

**Comment by whatyouhide** [comment]
@lexmag thanks for having a look, and you're very right :) Fixed in the last couple of commits. Let me know what you think!

File: lib/ex_unit/test/ex_unit/assertions_test.exs [hunk 2]
```diff
@@ -336,6 +336,38 @@ defmodule ExUnit.AssertionsTest do
...
   end
 
+  test "assert raise comparing messages (for equality)" do
+    assert_raise RuntimeError, "foo", fn ->
+      raise RuntimeError, "bar"
+    end
+  rescue
+    error in [ExUnit.AssertionError] ->
+      expected = String.rstrip("""
+      Wrong message for RuntimeError
```
Comment (lexmag) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: Maybe go using `<>` like in every other test? [For example](https://github.com/whatyouhide/elixir/blob/better-assert-raise-messages/lib/ex_unit/test/ex_unit/assertions_test.exs#L157-L161).

Comment (lexmag) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: Consistency, you know. :smile:

Comment (whatyouhide) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: Aaargh, yes you're right :) do you think `<>` would be better in the source as well or just the tests?

Comment (lexmag) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: I think `<>` would be preferable.

**Comment by whatyouhide** [comment]
@lexmag should be good to go!

**Comment by lexmag** [comment]
Nice. :clap:

File: lib/ex_unit/lib/ex_unit/assertions.ex [hunk 3]
```diff
@@ -392,8 +392,11 @@ defmodule ExUnit.Assertions do
       Regex.regex?(message) -> Exception.message(error) =~ message
     end
 
-    msg = "Wrong message for #{inspect exception}. " <>
-          "Expected #{inspect message}, got #{inspect Exception.message(error)}"
+    msg = "Wrong message for #{inspect exception}\n" <>
```
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/assertions.ex]: I think we need a trailing period here. as in the old one.

Comment (whatyouhide) [inline lib/ex_unit/lib/ex_unit/assertions.ex]: @eksperimental I initially put it there but removed it as there's a newline now. Other error messages don't have a trailing period. Wdyt?

File: lib/ex_unit/test/ex_unit/assertions_test.exs [hunk 4]
```diff
@@ -336,6 +336,32 @@ defmodule ExUnit.AssertionsTest do
...
+    assert_raise RuntimeError, "foo", fn ->
+      raise RuntimeError, "bar"
+    end
+  rescue
+    error in [ExUnit.AssertionError] ->
+      "Wrong message for RuntimeError" <>
+      "\nExpected:" <>
+      "\n  \"foo\"" <>
+      "\nGot:" <>
+      "\n  \"bar\"" = error.message
```
Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: It's confusing and not consistent the leading `\n`, in the other test in this commit you use a trailing `\n`

File: lib/ex_unit/test/ex_unit/assertions_test.exs [hunk 5]
```diff
@@ -336,6 +336,32 @@ defmodule ExUnit.AssertionsTest do
...
+    assert_raise RuntimeError, ~r/ba[zk]/, fn ->
+      raise RuntimeError, "bar"
+    end
+  rescue
+    error in [ExUnit.AssertionError] ->
+      "Wrong message for RuntimeError" <>
+      "\nExpected:" <>
+      "\n  ~r/ba[zk]/" <>
+      "\nGot:" <>
+      "\n  \"bar\"" = error.message
```
Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: same here about the leading `\n`

**Comment by eksperimental** [comment]
thank you very much @whatyouhide. I have added three comments. sorry, i coudln t make it before it was merged.

File: lib/ex_unit/test/ex_unit/assertions_test.exs (hunk 5 above)
Comment (whatyouhide) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: @eksperimental I prefer (by far) a trailing newline, but other tests in the same file use a leading one. Hence, I used a leading one in the tests and a trailing ones in the code :)

File: lib/ex_unit/test/ex_unit/assertions_test.exs (hunk 4 above)
Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: not this commit, but other test in general.

File: lib/ex_unit/lib/ex_unit/assertions.ex (hunk 3 above)
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/assertions.ex]: no dot is fine

File: lib/ex_unit/test/ex_unit/assertions_test.exs (hunk 5 above)
Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: is there any reason to use in the other tests the leading `\n`? i find it od and also difficult to visualize, like for example here, "Wrong Message"  is not lined up with "Expected",
compared to the example in the .ex code, where they look really neat

Comment (whatyouhide) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: I followed [this](https://github.com/elixir-lang/elixir/blob/master/lib/ex_unit/test/ex_unit/assertions_test.exs#L157-L161) lines in the same test file. That said, I'm seeing both styles are used [elsewhere](https://github.com/elixir-lang/elixir/blob/master/lib/ex_unit/test/ex_unit/assertions_test.exs#L122-L126) in that file. Anyways, I wouldn't worry so much about such a "style" issue. Another commit just to change this is history pollution IMO. What do you think?

Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: true, if it's just only for this issue in particular.
If we are ok with stardardizing the format, I can correct all of them, and submit one PR

Comment (whatyouhide) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: @eksperimental honestly, I would wait to do that. The reason is: I always get annoyed by heredocs having a trailing newline, and would love a way to avoid that. For example, a modifier for the `~s`/`~S` sigils like `n` (for *n*o*n*newlines) so that you can write:

```elixir
~s"""
this heredoc has no trailing newlines
"""n
```

I'll probably write that on the core mailing list, is it ok for you to wait a little while before making these changes?

Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: yeah man, go ahead please. just ping here for whatever resolution was taken, I will follow up through the mailing list away.

Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/assertions_test.exs]: Just to give a little bit more of context, 
besides the [lines that you took as reference](https://github.com/elixir-lang/elixir/blob/master/lib/ex_unit/test/ex_unit/assertions_test.exs#L157-L161), there is only one line in another test that uses that convention, and that's it for the tests.
https://gist.github.com/eksperimental/68403926802cb6a967da#file-leading_n-log-L2-L4

the rest of the cases are 3 .ex files using it, and in one of them the usage is proper, i think the other two it can be corrected.

File: lib/ex_unit/lib/ex_unit/assertions.ex (hunk 3 above)
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/assertions.ex]: @whatyouhide you (and @josevalim) were right. In Mix error messages that start with capital letter, we don't use a trailing period. I have fixed so many of those, that I forgot what the actual rule was.. haha
